### PR TITLE
Remove "Engine" notion

### LINKED ISSUE #391: [v2] Remove "Engine" notion
The notion of `Engine` is relatively abstract and pretty tied to `tartiflette` itself.

In the context of the development of a v2, could we not consider the removal of this notion of `Engine` in favor of the notion of `GraphQLSchema` and start on a more homogeneous solution with standards and existing solutions?

The `Engine` class is mainly used to abstract the `bake` process of the `GraphQLSchema` (loading of external modules, baking...) and then to allow to execute a request or a subscription according to the options provided at initialization.

However, the options provided to the `Engine` are either specific to `GraphQLSchema` or to the execution of a query which makes them options easily movable, for example:
```python
async def create_schema(
    sdl: Union[str, List[str]],
    schema_name: str = "default",
    custom_default_resolver: Optional[Callable] = None,
    custom_default_type_resolver: Optional[Callable] = None,
    modules: Optional[Union[str, List[str], List[Dict[str, Any]]]] = None,
    json_loader: Optional[Callable[[str], Dict[str, Any]]] = None,
    custom_default_arguments_coercer: Optional[Callable] = None,
) -> GraphQLSchema:
    pass


async def execute(
    schema: GraphQLSchema,
    query: Union[str, bytes],
    operation_name: Optional[str] = None,
    context: Optional[Any] = None,
    variables: Optional[Dict[str, Any]] = None,
    initial_value: Optional[Any] = None,
    error_coercer: Callable[[Exception, Dict[str, Any]], Dict[str, Any]] = None,
    query_cache_decorator: Optional[Callable] = UNDEFINED_VALUE,
) -> Dict[str, Any]:
    pass
```

The advantages:
* removal of an abstraction layer by necessarily necessary
* getting closer to the standards and other libraries

The inconvenients:
* significant break changes
* lookup how to deal with the `on_schema_execution` & `on_schema_subscription` directive hooks

### PR DESCRIPTION
**Pull request by Maximilien-R**
Closes #391.

As stated on the [related issue](https://github.com/tartiflette/tartiflette/issues/391), this PR aims to remove the "Engine" notion from `tartiflette` which is relatively abstract and which can lead to some bottleneck.

Here is my proposal to solve this issue:
* build schema through the `create_schema` utility function
* build `execute` function through the `executor_factory` utility function
* build `subscribe` function through the `subscribor_factory` utility function

Implementation POV, it would looks like this:
```python
# Initialize the GraphQL schema with the SDL
schema = await create_schema(
    "type Query { hello(name: String!) }",
    modules=["my_app.resolvers"],
    name="HelloWorld",
)

# Build the `execute` method for our schema
execute = executor_factory("HelloWorld")  # or executor_factory(schema)

# Execute a query
response = await execute("query HelloWorld { hello(name: "John") }")
```

Also, in addition to the features reported from the engine, two new parameters have been added:
* `parser`: which allows to specify a custom callable which will be used to parse the query and convert it into a `DocumentNode` (this can be use to provide a homemade parser or for example to use the one provided by `tartiflette` whith another JSON loads library than the builtin on)
* `rules`:  which allows to specify the list of validation rules to apply when validation the query `DocumentNode` (cf. #419)

Pros:
* removal of an abstraction layer by necessarily necessary
* getting closer to the standards and other libraries
* more flexible than `Engine` implementation
* better isolation and split of the notion of schema and executor/subscribor

Cons:
* significant break changes
* a little bit more verbose since you'll need to first create your schema and then your executor and/or subscribor while before you only had to create an engine instance

-----

NB: before going further, I would like to have input on this to be sure we are agree on the solution or if you have other idea for this. Also, we'll need to do a big refactoring on how we write/run our tests. Since we will remove the `Engine` notion, all our tests will fail and we'll need to write a lot of them 😢

### CODE REVIEW

File: tartiflette/execution/factory.py [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+            errors,
+            operation_name,
+            context,
+            variables,
+            initial_value,
+            context_coercer=context,
+        )
+
+
+class OperationSubscribor(BaseOperationPerformer):
```
Comment (abusi) [inline tartiflette/execution/factory.py:203]: `Subscriptor` or `Subscriber` :p

Comment (Maximilien-R) [reply tartiflette/execution/factory.py]: I renamed all occurrences to `Subscriptor` 😆

**Comment by Maximilien-R** [comment]
This PR should be ready for review now 👍 .

Here are the main things to know about it:
* introduce a `create_schema` function to build a GraphQLSchema
* introduce a `executor_factory` function to build an `execute` function linked to a GraphQLSchema
* introduce a `subscriptor_factory` function to build a `subscribe` function linked to a GraphQLSchema
* introduce a `create_schema_with_operationers` function to build a GraphQLSchema and its operationers (`execute` & `subscribe` function) at once
* get ride of the `Engine` class and occurrences along with its `create_engine` utility function
* introduce a new `pytest` marker to replace the `ttftt_engine` one which is now called `with_schema_stack`. The new marker allow to parametrize a `SchemaStack` which is an object containing a `GraphQLSchema` instance and its operationers. It can pop a `preset` schema (`pets`,  `libraries`, `coercion`, ...) or a custom one if the `sdl` paremeter is provided. Also, you can pass a `bakery` parameter which expect to have a callable with a `schema_name` attribute which then be called just before creating the `GraphQLSchema` instance to load resolvers, scalars, directives etc... as the `modules` attribute would do but without having to create a dedicated module for it. For more information on it, look at the different updated tests 👍 
* reworks tests (unit and functional) either to use `with_schema_stack` marker or to move all `conftest` into a single one usable for both unit and functional tests or to reunite all SQL into a package providing some utility function for them

File: tartiflette/__init__.py [hunk 2]
```diff
@@ -1,95 +1,96 @@
...
     "TartifletteError",
+    "create_schema",
+    "executor_factory",
+    "subscriptor_factory",
+    "create_schema_with_operationers",
 )
 
 
-async def create_engine(
+async def create_schema_with_operationers(
```
Comment (abusi) [inline tartiflette/__init__.py:26]: `create_schema_and_operators` seems to be a better name for me ? what do you think ?

File: tartiflette/execution/factory.py [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+from tartiflette.utils.callables import is_valid_coroutine
+from tartiflette.utils.errors import (
+    default_error_coercer,
+    error_coercer_factory,
+)
+
+__all__ = ("executor_factory", "subscriptor_factory")
+
+
+class BaseOperationer:
```
Comment (abusi) [inline tartiflette/execution/factory.py:24]: `Operator` is better than `Operationer` IMHO

File: tartiflette/schema/registry.py [hunk 4]
```diff
@@ -210,6 +210,17 @@ def find_schema_info(schema_name: str = "default") -> Dict[str, Any]:
         """
         return SchemaRegistry._schemas[schema_name]
 
+    @staticmethod
```
Comment (abusi) [inline tartiflette/schema/registry.py:213]: :+1:

File: tartiflette/schema/registry.py [hunk 5]
```diff
@@ -210,6 +210,17 @@ def find_schema_info(schema_name: str = "default") -> Dict[str, Any]:
...
+    def is_schema_registered(schema_name: str) -> bool:
```
Comment (abusi) [inline tartiflette/schema/registry.py:214]: :+1:

File: tests/conftest.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    SchemaRegistry.clean()
+
+
+@pytest.fixture
+def random_schema_name():
+    return uuid4().hex
+
+
+@pytest.fixture
+def random_schema_name():
```
Comment (abusi) [inline tests/conftest.py:46]: It seems to be defined twice :p

File: tests/conftest.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+def pytest_runtest_setup(item):
+    marker = _fetch_with_schema_stack_marker(item)
+    if not marker:
+        return
+
+    item.callspec.params[
+        "schema_stack"
+    ] = asyncio.get_event_loop().run_until_complete(
+        item.callspec.params["schema_stack"]()
+    )
```
Comment (abusi) [inline tests/conftest.py:135]: I really like this marker ! It's a very good job @Maximilien-R :heart:

File: tests/conftest.py (hunk 6 above)
Comment (Maximilien-R) [reply tests/conftest.py:46]: Oops, nice catch 👀

**Review by abusi — APPROVED** [review]

**Pull request merged**